pbm/restore/logical.go

```diff
@@ -253,11 +262,18 @@ func (r *Restore) Snapshot(
 		return err
 	}
 
-	// drop sharded dbs on sharded cluster, on each shard (not CSRS), only for full restore
-	if r.nodeInfo.IsSharded() && !r.nodeInfo.IsConfigSrv() && !util.IsSelective(nss) {
-		err = r.dropShardedDBs(ctx, bcp)
-		if err != nil {
-			return err
+	// drop databases on the sharded cluster as part of cleanup phase, on each shard (not CSRS)
+	if r.nodeInfo.IsSharded() && !r.nodeInfo.IsConfigSrv() {
+		if util.IsSelective(nss) {
+			err = r.selRestoreDBCleanup(ctx, nss)
+			if err != nil {
+				return errors.Wrap(err, "selective restore cleanup")
+			}
+		} else {
+			err = r.fullRestoreDBCleanup(ctx, bcp)
+			if err != nil {
+				return errors.Wrap(err, "full restore cleanup")
+			}
 		}
 	}
```

Comment:
I'd extract this into standalone method e.g. `r.shardedDBCleanup` or w/e.

IMHO it's complex enough set of conditions so we don't want to maintain it at two different places exactly the same